src/shogun/converter/StochasticProximityEmbedding.h

```diff
@@ -7,72 +7,74 @@
- * to form the second set.
- *
- * The parameter K for K-Nearest Neighbors in SPE_LOCAL corresponds to the class member
- * "m_k". Each of the two sets used on every iteration is formed by "m_nupdates" feature
- * vectors. Therefore, the number of feature vectors given must be always at least two
- * times the value of "m_nupdates".
- *
- * In order to avoid problems with memory in case a large number of features vectors is
- * to be embedded, the distance matrix is never computed explicitily. This has the
- * drawback that it is likely that the same distances are computed several times during
- * the process.
- *
- * Uses implementation from the Tapkee library.
- *
- * Only EuclideanDistance distance is supported for the moment.
- *
- */
-
-class StochasticProximityEmbedding : public EmbeddingConverter
-{
+	/** Stochastic Proximity Embedding (SPE) strategy */
+	enum ESPEStrategy
+	{
+		SPE_GLOBAL,
+		SPE_LOCAL,
+	};
+
+	/** @brief class StochasticProximityEmbedding used to construct embeddings
+	 * of data using the Stochastic Proximity algorithm.
+	 *
+	 * Agrafiotis, D. K. (2002)
+	 * Stochastic Proximity Embedding
+	 * Retrieved from:
+	 * http://www.dimitris-agrafiotis.com/Papers/jcc20078.pdf
+	 *
+	 * This class provides two different strategies for the computation of the
+	 * embedding. In each iteration, both strategies choose two sets of feature
+	 * vectors whose representation in the embedded space is updated. The first
+	 * set is randomly chosen in both strategies. However, the second set is
+	 * obtained differently depending on the strategy used. In the SPE_GLOBAL
+	 * strategy, the second set is still chosen at random. On the other hand, if
+	 * SPE_LOCAL is used, first of all, the K-Nearest Neighbors  of each of the
+	 * feature vectors in the first set is obtained and secondly, a number of
+	 * feature vectors among these K-Nearest Neighbors is chosen to form the
+	 * second set.
+	 *
+	 * The parameter K for K-Nearest Neighbors in SPE_LOCAL corresponds to the
+	 * class member "m_k". Each of the two sets used on every iteration is
+	 * formed by "m_num_updates" feature vectors. Therefore, the number of
+	 * feature vectors given must be always at least two times the value of
+	 * "m_num_updates".
+	 *
+	 * In order to avoid problems with memory in case a large number of features
+	 * vectors is to be embedded, the distance matrix is never computed
+	 * explicitly. This has the drawback that it is likely that the same
+	 * distances are computed several times during the process.
+	 *
+	 * Uses implementation from the Tapkee library.
+	 *
+	 * Only EuclideanDistance distance is supported for the moment.
+	 *
+	 */
+
+	class StochasticProximityEmbedding : public EmbeddingConverter
+	{
 
 	public:
-
 		/** constructor */
 		StochasticProximityEmbedding();
 
 		/** destructor */
-		virtual ~StochasticProximityEmbedding();
+		~StochasticProximityEmbedding() override;
```

Comment:
no virtual destructor?

Reply:
I follow https://en.cppreference.com/w/cpp/language/virtual#In_detail

Reply:
More here https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-override.html

Reply:
Hmm but this is a destructor. You are overriding the base class destructor so there will be leaks..

Reply:
"override" keyword simply means "this function is marked as virtual in some base class"

Reply:
Concluding everything, all StochasticProximityEmbedding base classes till SGObject (EmbeddingConverter, Converter, Transformer) should have the override keyword instead of virtual.

Reply:
Imagine we have a base class B having a method X and derived classes D_1, D_2, ..., D_N declared so: class D_1 : B; class D_2: D_1; ... 
The most correct way is to declare X as virtual in B and as override it in D_1, ..., D_N-1 and as final in D_N.

Reply:
Just had a look online. This override doesn’t do anything. The destructor is always virtual, because it needs to lookup the vtable along the hierarchy to know how to destruct all class members. The override only ensure that the base class is virtual, which is a given in the way shogun is designed from SGObject down. Imo this syntax is confusing and doesn’t bring anything. It says override, but it doesn’t, and to me at lest it looks like a bug (even though it actually isn’t).

Reply:
Btw I know how virtual functions work. My point was that overriding a destructor imo would imply overriding the behaviour of the parent classes. It doesn’t, because the destructor override here is a noop.

Reply:
https://github.com/isocpp/CppCoreGuidelines/issues/721

Reply:
> It says override, but it doesn’t

[C++11 Standart](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3690.pdf)
"Even though destructors are not inherited, a destructor in a derived class **_overrides_** a base class destructor." -- Page 233.
<br>
"The use of the _virtual_ specifier in the declaration of an overriding function is legal but redundant (has empty semantics)." -- Page 231.

Reply:
But in that case what's the point of `override`, if it overriden anyway? Other than to make sure that the base class has a virtual destructor, which we know it does? Anyway. up to you, my point is that then we need to change every destructor in shogun, otherwise we just end up with a mix and match of patterns and code styles.

Reply:
> Anyway. up to you, my point is that then we need to change every destructor in shogun

I can do it in an hour without any future problems.

>But in that case what's the point of override, if it overriden anyway?

C++ headers were invented to developers who use someone else code. When you first look at a header file, you do not know anything about the base class. As using _virtual_ specifier in the derived class has no sense, we should specify somehow that the method overrides the base method. Plus using override by analogy to writing _private_ specifier helps us not to shoot ourselves in the foot.

Reply:
https://github.com/chromium/chromium/search?q=override&unscoped_q=override
Chromium guys agree with me.

Reply:
go for it if you think that is worth it :)

Reply:
for reference I think @vinnik-dmitry07 means this https://github.com/chromium/chromium/blob/2ca8c5037021c9d2ecc00b787d58a31ed8fc8bcb/tools/clang/plugins/tests/virtual_specifiers.txt

Reply:
@vinnik-dmitry07 could you hold your horses for now and just follow shogun's style? We don't want to mix up ways we do things, as this leads to redundant discussions and compiler warnings. If you are keen in changing the way this is done in shogun, please open an issue and a separate PR that applies what you want to do to all of shogun.

Reply:
Of course such a PR is extremely welcome. (Although I suggest you defer this to later and focus on sending something related to your project proposal for now)